File: src/Scripts/Matlab/labview_functions.m
function [FUNC] = labview_functions()
FUNC.Console.Clear = @Console_Clear;
FUNC.Console.GetLines = @Console_GetLines;
FUNC.Console.Print = @Console_Print;

FUNC.TimingMethods.SendArray = @TimingMethods_SendArray;
FUNC.TimingMethods.SendComplexCluster = @TimingMethods_SendComplexCluster;
FUNC.TimingMethods.SendDBL = @TimingMethods_SendDBL;
FUNC.TimingMethods.SendVarious = @TimingMethods_SendVarious;

function  Console_Clear (Con) 
% Clear the console of all lines

% input data types
% output data types

RemoteCall = remotecall();
parIn=struct;
[err, parOut] = RemoteCall.Execute(Con, "Console", "Clear", parIn);
if err.status == 1, error (err.source); end

end

function [Lines] = Console_GetLines (Con) 
% Get all the lines currently on the console

% input data types
% output data types
% Lines = jsondecode("""""");

RemoteCall = remotecall();
parIn=struct;
[err, parOut] = RemoteCall.Execute(Con, "Console", "GetLines", parIn);
if err.status == 1, error (err.source); end
Lines = parOut.Lines;

end

function  Console_Print (Con, Line)
% Print a line to the console

% input data types
% Line = jsondecode("""""");
% output data types

RemoteCall = remotecall();
parIn={};
parIn.Line = Line;

[err, parOut] = RemoteCall.Execute(Con, "Console", "Print", parIn);
if err.status == 1, error (err.source); end

end

function [Out] = TimingMethods_SendArray (Con, In)

% input data types
% In = jsondecode("[]");
% output data types
% Out = jsondecode("[]");

RemoteCall = remotecall();
parIn={};
parIn.In = In;

[err, parOut] = RemoteCall.Execute(Con, "TimingMethods", "SendArray", parIn);
if err.status == 1, error (err.source); end
Out = parOut.Out;

end

function [Out] = TimingMethods_SendComplexCluster (Con, In)

% input data types
% In = jsondecode("{""subcl"":{""dbl"":[]},""str"":[],""Array"":[],""Path"":""""}");
% output data types
% Out = jsondecode("{""subcl"":{""dbl"":[]},""str"":[],""Array"":[],""Path"":""""}");

RemoteCall = remotecall();
parIn={};
parIn.In = In;

[err, parOut] = RemoteCall.Execute(Con, "TimingMethods", "SendComplexCluster", parIn);
if err.status == 1, error (err.source); end
Out = parOut.Out;

end

function [Out] = TimingMethods_SendDBL (Con, In)

% input data types
% In = jsondecode("0.00000000000000");
% output data types
% Out = jsondecode("0.00000000000000");

RemoteCall = remotecall();
parIn={};
parIn.In = In;

[err, parOut] = RemoteCall.Execute(Con, "TimingMethods", "SendDBL", parIn);
if err.status == 1, error (err.source); end
Out = parOut.Out;

end

function [Array_Out, Cluster_Out, Boolean_Out] = TimingMethods_SendVarious (Con, Boolean_In, Array_in, Cluster_In)

% input data types
% Boolean_In = jsondecode("true");
% Array_in = jsondecode("[]");
% Cluster_In = jsondecode("{""subcl"":{""dbl"":[]},""str"":[],""Array"":[],""Path"":""""}");
% output data types
% Array_Out = jsondecode("[]");
% Cluster_Out = jsondecode("{""subcl"":{""dbl"":[]},""str"":[],""Array"":[],""Path"":""""}");
% Boolean_Out = jsondecode("true");

RemoteCall = remotecall();
parIn={};
parIn.Boolean_In = Boolean_In;
parIn.Array_in = Array_in;
parIn.Cluster_In = Cluster_In;

[err, parOut] = RemoteCall.Execute(Con, "TimingMethods", "SendVarious", parIn);
if err.status == 1, error (err.source); end
Array_Out = parOut.Array_Out;
Cluster_Out = parOut.Cluster_Out;
Boolean_Out = parOut.Boolean_Out;

end


end

File: src/Scripts/Python/labview_functions.py
import json
import zmq

class RemoteCall:
    """Class for connecting and executing commands over ZeroMQ Using JSON
    """
    socket = 0

    def connect(self,connection_string):
        """ Make the initial connection to the client running in LabVIEW

        Parameters
        ----------
        connection_string : String (Example "tcp://127.0.0.1:5555", "ConnectionType"://"IPAdress":"Port")
        """

        context = zmq.Context()
        self.socket = context.socket(zmq.REQ)
        self.socket.connect(connection_string)

    def execute(self, class_name, method, par_in):
        """ Execute the Method by serializing inputs sending it over, and deserializing the result
        """
        data_2_send = "{ \"Command\":\"Execute\" }\x17{\"ClassLabel\":\"%s\",\"Method\":\"%s\"}\x17%s" % (class_name, method, json.dumps({"In": par_in}))
        self.socket.send_string (data_2_send)
        message = self.socket.recv()
        parts = message.split(chr(0x17).encode())
        return json.loads(parts[1]), json.loads(parts[0])

class LabVIEWerror(Exception):
    """Base class for all LabVIEW errors

    Attributes
    ----------
    code: error code for testing which error happend
    source: source of the error
    """
    def __init__(self, *args: object) -> None:
        self.code = self.args[0]
        self.source = self.args[1]
        super().__init__(*args)

    def __str__(self) -> str:
        return f"LabVIEW Error with code: {self.code} source: {self.source}"

class LabviewControl:
    """ Class for calling all labview function with connection embedded
    """

    def __init__(self,connection_string) -> None:
        self._connection=RemoteCall()
        self._connection.connect(connection_string)
        self.Console=Console(self._connection)
        self.PerformanceTest=PerformanceTest(self._connection)

class Console:
    """Class for running a small console in labview with a display showing printed lines
    """
    def __init__(self, con) -> None:
        self._connection=con

    def Clear (self):
        """Clear the console of all lines

        Parameters
        ----------

        Returns
        ----------

        """

        par_in = {
        }

        err, par_out = self._connection.execute("Console", "Clear", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])

    def GetLines (self):
        """Get all the lines currently on the console

        Parameters
        ----------

        Returns
        ----------
        Lines : String

        """

        par_in = {
        }

        err, par_out = self._connection.execute("Console", "GetLines", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])
        return par_out["Lines"]

    def Print (self, Line):
        """Print a line to the console

        Parameters
        ----------
        Line : String

        Returns
        ----------

        """

        par_in = {
        "Line" : Line
        }

        err, par_out = self._connection.execute("Console", "Print", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])

class PerformanceTest:
    """Class for testing timing, data types and performance of sending data between LabVIEW and the calling language
    """
    def __init__(self, con) -> None:
        self._connection=con

    def SendArray (self, In):
        """

        Parameters
        ----------
        In : List [Float]

        Returns
        ----------
        Out : List [Float]

        """

        par_in = {
        "In" : In
        }

        err, par_out = self._connection.execute("PerformanceTest", "SendArray", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])
        return par_out["Out"]

    def SendComplexCluster (self, In):
        """

        Parameters
        ----------
        In : Dict {"subcl" : Dict {"dbl" : List [Float]}, "str" : List [String], "Array" : List [Dict {"Boolean" : Bool, "String" : String, "Array 2" : List [Float]}], "Path" : Unsupported datatype Path}

        Returns
        ----------
        Out : Dict {"subcl" : Dict {"dbl" : List [Float]}, "str" : List [String], "Array" : List [Dict {"Boolean" : Bool, "String" : String, "Array 2" : List [Float]}], "Path" : Unsupported datatype Path}

        """

        par_in = {
        "In" : In
        }

        err, par_out = self._connection.execute("PerformanceTest", "SendComplexCluster", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])
        return par_out["Out"]

    def SendDBL (self, In):
        """

        Parameters
        ----------
        In : Float, Control for sending a double

        Returns
        ----------
        Out : Float, Indicator for value filled by the control at the input. Data should be the same.

        """

        par_in = {
        "In" : In
        }

        err, par_out = self._connection.execute("PerformanceTest", "SendDBL", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])
        return par_out["Out"]

    def SendVarious (self, Boolean_In, Array_in, Cluster_In):
        """

        Parameters
        ----------
        Boolean_In : Bool,
        Array_in : List [Int],
        Cluster_In : Dict {"subcl" : Dict {"dbl" : List [Float]}, "str" : List [String], "Array" : List [Dict {"Boolean" : Bool, "String" : String, "Array 2" : List [Float]}], "Path" : Unsupported datatype Path}

        Returns
        ----------
        Array_Out : List [Int],
        Cluster_Out : Dict {"subcl" : Dict {"dbl" : List [Float]}, "str" : List [String], "Array" : List [Dict {"Boolean" : Bool, "String" : String, "Array 2" : List [Float]}], "Path" : Unsupported datatype Path},
        Boolean_Out : Bool

        """

        par_in = {
        "Boolean_In" : Boolean_In,
        "Array_in" : Array_in,
        "Cluster_In" : Cluster_In
        }

        err, par_out = self._connection.execute("PerformanceTest", "SendVarious", par_in)
        if err["status"]:
            raise LabVIEWerror(err["code"],err["source"])
        return par_out

